kotlin-native/backend.native/tests/objcexport/refined.kt

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2010-2022 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package refined
+
+@Target(AnnotationTarget.PROPERTY, AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@ObjCRefined
+annotation class MyObjCRefined
+
+@Target(AnnotationTarget.PROPERTY, AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@SwiftRefined
+annotation class MySwiftRefined
+
+@ObjCRefined
+fun foo(): Int = 1
+
+@SwiftRefined
+fun fooRefined(): String = foo().toString()
+
+@MyObjCRefined
+fun myFoo(): Int = 2
+
+@MySwiftRefined
+fun myFooRefined(): String = myFoo().toString()
+
+@ObjCRefined
+val bar: Int = 3
+
+@SwiftRefined
+val barRefined: String get() = bar.toString()
+
+@MyObjCRefined
+val myBar: Int = 4
+
+@MySwiftRefined
+val myBarRefined: String get() = myBar.toString()
```

Comment:
> A property might have any of its accessors annotated.
> A class is nested in an annotated class.

Those cases aren't allowed, the annotations are only allowed on properties or functions.
Note: these restrictions should be enforced by the compiler for custom refinement annotations as well.

> In case of override, when one of the methods is annotated while the other one is not.

This shouldn't be allowed either.

Reply:
> Those cases aren't allowed, the annotations are only allowed on properties or functions.

Right! Sorry.

> Note: these restrictions should be enforced by the compiler for custom refinement annotations as well.

Oh. What if a compiler/KSP plugin can work with e.g. classes, but the annotation that triggers this plugin is annotated with `@*Refined`? In this case, the compiler won't allow using it on classes.

Like, the annotation is 
```kotlin
@ObjCRefined
annotation class Beautify
```
It instructs the plugin to provide some beautified wrappers for the annotated APIs. And can be applied to the class -- in this case, it e.g. affects all the members. Should we prohibit that?

The workaround would probably be like "the plugin provides specific annotations for classes" (like `BeautifyMembers`). Are we ok with that?

Reply:
> It instructs the plugin to provide some beautified wrappers for the annotated APIs. And can be applied to the class -- in this case, it e.g. affects all the members. Should we prohibit that?

The frontend-checker current prohibits this.

> The workaround would probably be like "the plugin provides specific annotations for classes" (like BeautifyMembers). Are we ok with that?

Well it isn't ideal, but I don't think there is a better solution.
If we would allow the `Beautify` annotation from your example then what should happen to its members?
Should all members be refined? The plugin would probably only like to refine specific type of members (e.g. with a specific return type).

Reply:
> Well it isn't ideal, but I don't think there is a better solution.
If we would allow the `Beautify` annotation from your example then what should happen to its members?
Should all members be refined? The plugin would probably only like to refine specific type of members (e.g. with a specific return type).

Ok.

> > In case of override, when one of the methods is annotated while the other one is not.

> This shouldn't be allowed either.

Is it fully prohibited though? https://github.com/JetBrains/kotlin/pull/4818/files#diff-0c432a0f5175969cadff3ccfb46ea94201b7f707b5004a23aed4377fe8dcd515R101

Reply:
Correct `barB` and `fooA` cover the case where only some of the overridden methods are annotated.

Reply:
Is there a test in this file (`refined.kt`) that checks this case -- when a non-annotated method overrides an annotated one?
`compiler/testData/diagnostics/nativeTests/objCRefinement.kt` test doesn't help because it doesn't check ObjCExport.